web/courses.go

```diff
@@ -184,88 +184,83 @@ func (s *QuickFeedService) getAllCourseSubmissions(request *qf.SubmissionRequest
 		return nil, err
 	}
 
-	var enrolLinks []*qf.EnrollmentLink
+	var submissionsMap map[uint64]*qf.Submissions
 	switch request.GetType() {
 	case qf.SubmissionRequest_GROUP:
-		enrolLinks = makeGroupResults(course, assignments)
+		submissionsMap = makeGroupResults(course, submissions, course.GetAssignments())
 	case qf.SubmissionRequest_USER:
-		enrolLinks = makeIndividualResults(course, assignments)
+		submissionsMap = makeIndividualResults(course, submissions, course.GetAssignments())
 	case qf.SubmissionRequest_ALL:
-		enrolLinks = makeAllResults(course, assignments)
+		submissionsMap = makeAllResults(course, submissions, course.GetAssignments())
 	}
-	return &qf.CourseSubmissions{Course: course, Links: enrolLinks}, nil
+	return &qf.CourseSubmissions{Submissions: submissionsMap}, nil
 }
 
 // makeGroupResults generates enrollment to assignment to submissions links
 // for all course groups and all group assignments
-func makeGroupResults(course *qf.Course, assignments []*qf.Assignment) []*qf.EnrollmentLink {
-	enrolLinks := make([]*qf.EnrollmentLink, 0)
+func makeGroupResults(course *qf.Course, submissions []*qf.Submission, assignments []*qf.Assignment) map[uint64]*qf.Submissions {
+	submissionsMap := make(map[uint64]*qf.Submissions)
 	seenGroup := make(map[uint64]bool)
+	om := newOrderMap(course.GetAssignments())
 	for _, enrollment := range course.Enrollments {
-		if seenGroup[enrollment.GroupID] {
+		if seenGroup[enrollment.GroupID] || enrollment.GroupID == 0 {
```

Comment:
This could be made simpler with:
```go
skipGroup := map[uint64]bool{
	0: true,  // ignore enrollments without a group
}
```
And then do:
```go
if skipGroup[enrollment.GroupID] {
	continue // include group enrollment only once
}
skipGroup[enrollment.GroupID] = true
```